### PR TITLE
[Operator] Add repeat_interleave_tensor

### PR DESCRIPTION
**Pull request by zfu82**
### Performance

Tested on NV-A100
```
benchmark/test_pointwise_perf.py::test_perf_repeat_interleave_tensor 
Operator repeat_interleave_tensor Performance Test (dtype=torch.int32, mode=cuda)
Size    Torch Latency (ms)    Gems Latency (ms)    Gems Speedup
---------------------------------------------------------------
1024              0.095232             0.259072           0.368
6144              0.094208             0.277504           0.339
11264             0.093184             0.279552           0.333
16384             0.095232             0.278528           0.342
21504             0.100352             0.278528            0.36
26624               0.1024             0.280576           0.365
31744             0.103424               0.2816           0.367
36864             0.103424             0.285696           0.362
41984             0.103424              0.28672           0.361
47104             0.104448             0.285696           0.366
52224             0.104448             0.287744           0.363
57344             0.104448             0.287744           0.363
62464             0.104448             0.288768           0.362
67584             0.104448             0.295936           0.353
72704             0.104448             0.297984           0.351
77824             0.106496             0.304128            0.35
PASSED
```

### CODE REVIEW

File: src/flag_gems/ops/repeat_interleave.py [hunk 1]
```diff
@@ -53,3 +57,41 @@ def repeat_interleave_self_int(inp, repeats, dim=None, *, output_size=None):
...
+    result_size = cumsum[-1].item()
+
+    assert cumsum[0].item() >= 0, "repeats can not be negative"
+    assert result_size >= 0, "repeats can not be negative"
+
+    out = torch.zeros((result_size,), dtype=repeats.dtype, device=repeats.device)
+    size = repeats.size(0)
+    inputs = torch.arange(0, size, dtype=repeats.dtype, device=repeats.device)
+
+    grid = (size - 1,)
```
Comment (iclementine) [inline src/flag_gems/ops/repeat_interleave.py]: This task partitioning along `repeats.size`(also the input's size at the repeat axis) may be in efficient when the repeat number is inbalanced.

I think we can consider a better task partitioning along the output's size at the repeat axis.

For example:

```text
Input: [a b c d]
repeats: [2 2 4 5]

Output: [a a b b c c c c d d d d d]
```

Instead of considering how many times and where to copy the input, we can consider the inverse, for each output, where to copy from the input.

```text
Input: [a b c d]
repeats: [2 2 4 5]

Output: [a a b b c c c c d d d d d]
Input Indices: [0 0 1 1 2 2 2 2 3 3 3 3 3]
```

We can generate the indices by comparing output indices with the cumsum of repeats and summing them.

|     |     |     |     |      |     |
| --- | --- | --- | --- | ---- | --- |
|     | >=2 | >=4 | 8>= | 13>= | sum |
| 0   | 0   | 0   | 0   | 0    | 0   |
| 1   | 0   | 0   | 0   | 0    | 0   |
| 2   | 1   | 0   | 0   | 0    | 1   |
| 3   | 1   | 0   | 0   | 0    | 1   |
| 4   | 1   | 1   | 0   | 0    | 2   |
| 5   | 1   | 1   | 0   | 0    | 2   |
| 6   | 1   | 1   | 0   | 0    | 2   |
| 7   | 1   | 1   | 0   | 0    | 2   |
| 8   | 1   | 1   | 1   | 0    | 3   |
| 9   | 1   | 1   | 1   | 0    | 3   |
| 10  | 1   | 1   | 1   | 0    | 3   |
| 11  | 1   | 1   | 1   | 0    | 3   |
| 12  | 1   | 1   | 1   | 0    | 3   |

With this input indices, we can easily parallelize the copying.

Comment (iclementine) [reply src/flag_gems/ops/repeat_interleave.py]: Oh I see, repeat_interleave.Tensor returns that indices only without copying the input `[a,b,c,d]`.

Since triton does not allow fine-grained task partitioning on warps or threads, compare and reduce may be a more efficient way.

Comment (zfu82) [reply src/flag_gems/ops/repeat_interleave.py]: Tested when `repeats = torch.randint(0, 30, (1024, ), dtype=torch.int32, device="cuda")`, I've got the following results:
torch: 0.0993 ms
pr: 0.1781 ms
compare and reduce: 0.3164 ms

File: src/flag_gems/ops/repeat_interleave.py [hunk 2]
```diff
@@ -53,3 +57,41 @@ def repeat_interleave_self_int(inp, repeats, dim=None, *, output_size=None):
...
+
+    cumsum = repeats.cumsum(axis=0)
+    result_size = cumsum[-1].item()
+
+    assert cumsum[0].item() >= 0, "repeats can not be negative"
+    assert result_size >= 0, "repeats can not be negative"
+
+    out = torch.zeros((result_size,), dtype=repeats.dtype, device=repeats.device)
+    size = repeats.size(0)
+    inputs = torch.arange(0, size, dtype=repeats.dtype, device=repeats.device)
```
Comment (iclementine) [inline src/flag_gems/ops/repeat_interleave.py]: Since input is initialized as arange(0, size), we can ommit it by using pid, and we can save a load from global memory.

File: src/flag_gems/ops/repeat_interleave.py [hunk 3]
```diff
@@ -53,3 +57,41 @@ def repeat_interleave_self_int(inp, repeats, dim=None, *, output_size=None):
...
+    inputs = tl.load(inputs_ptr + pid + 1, mask)
+    repeats = tl.load(repeats_ptr + pid + 1, mask)
+
+    tl.device_assert(repeats >= 0, "repeats can not be negative")
+
+    out_offsets = cumsum
+    for _ in range(repeats):
+        tl.store(out_ptr + out_offsets, inputs)
+        out_offsets += 1
+
```
Comment (iclementine) [inline src/flag_gems/ops/repeat_interleave.py]: Maybe write out by tile instead of by element.

something like

```python
    out_ptr += out_offset
    for start_k in range(0, repeats, 32):
        offsets_k = start_k + tl.arange(0, 32)
        mask_k = offsets_k < repeats
        tl.store(out_ptr + offsets_k, pid, mask=mask_k)
```

File: src/flag_gems/ops/repeat_interleave.py (hunk 1 above)
Comment (iclementine) [reply src/flag_gems/ops/repeat_interleave.py]: Oh, it is not as good as expected.

File: src/flag_gems/ops/repeat_interleave.py (hunk 2 above)
Comment (zfu82) [reply src/flag_gems/ops/repeat_interleave.py]: OK

File: src/flag_gems/ops/repeat_interleave.py (hunk 3 above)
Comment (zfu82) [reply src/flag_gems/ops/repeat_interleave.py]: Done. Retesting the performance on NV-A100, compared to the previous implementation, we've made some improvements.

```
benchmark/test_pointwise_perf.py::test_perf_repeat_interleave_tensor 
Operator repeat_interleave_tensor Performance Test (dtype=torch.int32, mode=cuda)
Size    Torch Latency (ms)    Gems Latency (ms)    Gems Speedup
---------------------------------------------------------------
1024              0.095232             0.139264           0.684
6144              0.093184             0.157696           0.591
11264             0.093184              0.15872           0.587
16384             0.095232              0.15872             0.6
21504             0.100352             0.159744           0.628
26624               0.1024             0.159744           0.641
31744             0.104448             0.161792           0.646
36864             0.104448             0.164864           0.634
41984             0.104448              0.16384           0.638
47104             0.104448             0.164864           0.634
52224             0.104448             0.165888            0.63
57344             0.104448             0.165888            0.63
62464             0.104448              0.17408             0.6
67584             0.105472             0.176128           0.599
72704             0.105472             0.181248           0.582
77824              0.10752             0.182272            0.59
PASSED
```
